src/Pirouette/Transformations/DCE.hs

```diff
@@ -10,19 +10,36 @@ import Data.Generics.Uniplate.Data
 import qualified Data.List as L
 import qualified Data.Map as M
 import qualified Data.Set as S
+import qualified Data.Text as T
 import Pirouette.Monad
 import Pirouette.Term.Syntax
 import qualified Pirouette.Term.Syntax.SystemF as SystF
 import Pirouette.Transformations.Utils
 
 data CtorRemoveInfo = CtorRemoveInfo
-  { criDtorName :: Name
+  { criCtorTypeName :: Name
+  , criCtorName :: Name
   , criCtorIdx :: Int
-  , criCtorTypeName :: Name
+  , criDtorName :: Name
   } deriving (Show, Eq, Ord)
 
-removeDeadCtors :: forall lang. (Language lang) => PrtUnorderedDefs lang -> PrtUnorderedDefs lang
-removeDeadCtors defs = L.foldl' (flip removeCtor) defs (findDeadCtors defs)
+newtype RemoveDeadCtorsOpts = RemoveDeadCtorsOpts
+  { rdcoToKeep :: M.Map T.Text [T.Text]
+  } deriving (Show)
+
+removeDeadCtors :: forall lang. (Language lang)
+                => RemoveDeadCtorsOpts
+                -> PrtUnorderedDefs lang
+                -> PrtUnorderedDefs lang
+removeDeadCtors RemoveDeadCtorsOpts{..} defs = L.foldl' (flip removeCtor) defs ctorsToRemove
+  where
+    ctorsToRemove = filter (not . shallKeep) $ findDeadCtors defs
```

Comment:
Sure, I can change it to have this behaviour. Indeed, it's not obvious what is more natural — the usr specifying a "to-keep list" (as a couple of comments above), or the explicit "to-remove" list.

> Which still suffers the problem of: what if the user forgets a constructor that is important to be kept?

Yep, even though mentioning a set of constructors to keep is isomorphic to the set of constructors to remove for a fixed type, a "to-remove" list is more future-proof in case the user adds more constructors later and forgets to update this list.